### PR TITLE
Split features into common and OpenGL features

### PR DESCRIPTION
**Pull request by mbouron-gpsw**

### CODE REVIEW

File: libnodegl/backends/gl/gpu_ctx_gl.c [hunk 1]
```diff
@@ -409,6 +409,39 @@ static void NGLI_GL_APIENTRY gl_debug_message_callback(GLenum source,
...
+static void gpu_ctx_info_init(struct gpu_ctx *s)
+{
+    const struct gpu_ctx_gl *s_priv = (struct gpu_ctx_gl *)s;
+    const struct glcontext *gl = s_priv->glcontext;
+
+    s->version = gl->version;
+    s->language_version = gl->glsl_version;
+    for (int i = 0; i < NGLI_ARRAY_NB(feature_map); i++) {
+        const int64_t feature = feature_map[i].feature;
+        const int64_t feature_gl = feature_map[i].feature_gl;
```
Comment (cboesch-gpsw) [inline libnodegl/backends/gl/gpu_ctx_gl.c:441]: `uint64_t` for consistency?

Also the intermediate `feature` looks unnecessary, but you can keep it if you prefer.

File: libnodegl/pgcraft.c [hunk 2]
```diff
@@ -1129,6 +1133,8 @@ static void setup_glsl_info_gl(struct pgcraft *s)
     struct ngl_ctx *ctx = s->ctx;
     const struct ngl_config *config = &ctx->config;
     struct gpu_ctx *gpu_ctx = ctx->gpu_ctx;
+    struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
+    struct glcontext *gl = gpu_ctx_gl->glcontext;
```
Comment (cboesch-gpsw) [inline libnodegl/pgcraft.c:1137]: could be const

File: libnodegl/vaapi_ctx.c [hunk 3]
```diff
@@ -38,16 +42,28 @@
 #include "internal.h"
 #include "vaapi_ctx.h"
 
+static int check_extensions(struct gpu_ctx *gpu_ctx)
+{
+#if defined(BACKEND_GL)
+    struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
+    struct glcontext *gl = gpu_ctx_gl->glcontext;
```
Comment (cboesch-gpsw) [inline libnodegl/vaapi_ctx.c:49]: could be const

File: tests/meson.build [hunk 4]
```diff
@@ -267,10 +268,15 @@ foreach backend : backends
     tests_rtt += [
       'feature_depth_msaa',
       'feature_depth_stencil_msaa',
+    ]
+
+    if has_ds_resolve
+      tests_rtt += [
```
Comment (cboesch-gpsw) [inline tests/meson.build:274]: please add a re-indent commit

File: libnodegl/backends/gl/gpu_ctx_gl.c [hunk 5]
```diff
@@ -425,6 +425,8 @@ static const struct {
     {NGLI_FEATURE_UNIFORM_BUFFER,               NGLI_FEATURE_GL_UNIFORM_BUFFER_OBJECT},
     {NGLI_FEATURE_STORAGE_BUFFER,               NGLI_FEATURE_GL_SHADER_STORAGE_BUFFER_OBJECT},
     {NGLI_FEATURE_DEPTH_STENCIL_RESOLVE,        NGLI_FEATURE_GL_FRAMEBUFFER_OBJECT},
+    {NGLI_FEATURE_TEXTURE_FLOAT_RENDERABLE,     NGLI_FEATURE_GL_COLOR_BUFFER_FLOAT},
+    {NGLI_FEATURE_TEXTURE_HALF_FLOAT_RENDERABLE,NGLI_FEATURE_GL_COLOR_BUFFER_HALF_FLOAT},
```
Comment (cboesch-gpsw) [inline libnodegl/backends/gl/gpu_ctx_gl.c:429]: can you elaborate in the commit description on the decision to name it "renderable" vs "color buffer"?

**Review by cboesch-gpsw — CHANGES_REQUESTED** [review]
Aside from cosmetics, LGTM

File: libnodegl/android_ctx.c [hunk 6]
```diff
@@ -97,11 +103,17 @@ int ngli_android_ctx_init(struct gpu_ctx *gpu_ctx, struct android_ctx *s)
...
 
+#if defined(BACKEND_GL)
     const struct ngl_config *config = &gpu_ctx->config;
-    const int features = NGLI_FEATURE_OES_EGL_EXTERNAL_IMAGE |
-                         NGLI_FEATURE_EGL_ANDROID_GET_IMAGE_NATIVE_CLIENT_BUFFER;
-    if (config->backend == NGL_BACKEND_OPENGLES && (gpu_ctx->features & features) == features)
-        s->has_native_imagereader_api = 1;
+    if (config->backend == NGL_BACKEND_OPENGLES) {
+        struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
+        struct glcontext *gl = gpu_ctx_gl->glcontext;
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/android_ctx.c:110]: Done.

File: libnodegl/backends/gl/gpu_ctx_gl.c (hunk 1 above)
Comment (mbouron-gpsw) [reply libnodegl/backends/gl/gpu_ctx_gl.c:441]: Done.

File: libnodegl/api.c [hunk 7]
```diff
@@ -481,6 +481,7 @@ static const char *get_cap_string_id(unsigned cap_id)
     case NGL_CAP_TEXTURE_3D:                    return "texture_3d";
     case NGL_CAP_TEXTURE_CUBE:                  return "texture_cube";
     case NGL_CAP_UINT_UNIFORMS:                 return "uint_uniforms";
+    case NGL_CAP_DEPTH_STENCIL_RESOLVE:         return "depth_stencil_resolve";
```
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/api.c]: Done in the relevant places.

File: tests/meson.build (hunk 4 above)
Comment (mbouron-gpsw) [reply tests/meson.build:274]: Done the re-indent directly in the commit.

File: libnodegl/android_ctx.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/android_ctx.c]: Done.

File: libnodegl/backends/gl/gpu_ctx_gl.c (hunk 5 above)
Comment (mbouron-gpsw) [reply libnodegl/backends/gl/gpu_ctx_gl.c:429]: Done.

File: libnodegl/pgcraft.c (hunk 2 above)
Comment (mbouron-gpsw) [reply libnodegl/pgcraft.c]: Done.

File: libnodegl/vaapi_ctx.c (hunk 3 above)
Comment (mbouron-gpsw) [reply libnodegl/vaapi_ctx.c]: Done.

File: libnodegl/api.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (cboesch-gpsw) [reply libnodegl/api.c]: no :(

**Review by cboesch-gpsw — APPROVED** [review]

File: libnodegl/api.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbouron-gpsw) [reply libnodegl/api.c]: Faulty brain sorry, should be OK now.

**Pull request merged**